File: node/report.js
#!/usr/bin/env node

var fs = require('fs')
var url = require('url')
var os = require('os')
var path = require('path')

function makeRequest (target, callback) {
  if (!/https?:/.test(target)) {
    target = 'http://' + target
  }

  var uri = url.parse(target)
  var http = require(uri.protocol === 'https:' ? 'https' : 'http')
  var params = {
    method: 'POST',
    hostname: uri.hostname,
    port: uri.port,
    path: uri.path,
    headers: {
      'Content-Type': 'application/json',
    },
  }

  return http.request(params, callback)
}

function body (tests) {
  if (process.env.TRAVIS === 'true') {
    return travisBody(tests)
  }

  if (process.env.CIRCLECI === 'true') {
    return circleBody(tests)
  }

  throw new Error('Build environment not supported')
}

function travisBody (tests) {
  return {
    authorization: process.env.SHELFGAUGE_AUTH,
    data: {
      ref: process.env.TRAVIS_COMMIT,
      name: process.env.TRAVIS_BRANCH,
      ranAt: new Date().toISOString(),
      tests: tests,
      env: {
        source: 'travis',
        info: [
          'CPU: ' + os.cpus()[0].model,
          'Platform: ' + os.platform(),
          'Build id: ' + process.env.TRAVIS_BUILD_ID,
          'Job id: ' + process.env.TRAVIS_JOB_ID
        ].join('\n')
      },
    },
  }
}

function circleBody (tests) {
  return {
    authorization: process.env.SHELFGAUGE_AUTH,
    data: {
      ref: process.env.CIRCLE_SHA1,
      name: process.env.CIRCLE_BRANCH,
      ranAt: new Date().toISOString(),
      tests: tests,
      env: {
        source: 'circle',
        info: [
          'CPU: ' + os.cpus()[0].model,
          'Platform: ' + os.platform(),
          'Build num: ' + process.env.CIRCLE_BUILD_NUM
        ].join('\n')
      },
    },
  }
}

function once (callback) {
  var wasCalled = false
  return function () {
    if (wasCalled) {
      return
    }

    callback.apply(this, arguments)
    wasCalled = true
  }
}

function post (tests, url, done) {
  var doneOnce = once(done)

  try {
    var data = JSON.stringify(body(tests))
    var req = makeRequest(url, function (res) {
      var data = ''
      res.on('data', function (chunk) {
        data += chunk
      })
      res.on('error', function (err) {
        doneOnce(err)
      })
      res.on('end', function () {
        doneOnce(null, data)
      })
    })

    req.on('error', function (err) {
      doneOnce(err)
    })

    req.write(data)
    req.end()
  } catch (err) {
    doneOnce(err)
  }
}

if (require.main === module) {
  var done = once(function (err, data) {
    if (err) {
      console.error(err)
      process.exit(1)
    } else {
      console.log(data)
    }
  })

  var file = process.argv[2]
  var addr = process.argv[3]

  if (!file || !addr) {
    return done("usage: node " + path.basename(__filename) + " tests.json http://shelfgauge.url")
  }

  try {
    var tests = JSON.parse(fs.readFileSync(file))
    post(tests, addr, done)
  } catch (err) {
    done(err)
  }
} else {
  exports.post = post
  exports.body = body
}

File: ruby/report.rb
#!/usr/bin/env ruby

require 'date'
require 'net/http'
require 'json'

module ShelfgaugeReport
  def self.post(tests, url, &block)
    if url !~ /https?:/
      url = 'http://' + url
    end

    uri = URI(url)
    data = JSON.dump(body(tests))

    Net::HTTP.start(uri.host.gsub(/^\[(.*)\]$/, '\1'), uri.port) do |http|
      path = uri.path.empty? ? '/' : uri.path
      http.request_post(path, data, "Content-Type" => "application/json") do |res|
        res.read_body(&block)
      end
    end
  end

  def self.body(tests)
    if (ENV['TRAVIS'] == 'true')
      return travis_body(tests)
    end

    throw new Error('Build environment not supported')
  end

  def self.travis_body(tests)
    compacterize({
      authorization: ENV['SHELFGAUGE_AUTH'],
      data: {
        ref: ENV['TRAVIS_COMMIT'],
        name: ENV['TRAVIS_BRANCH'],
        pullRequest: env('TRAVIS_PULL_REQUEST', ignore: 'false'),
        ranAt: DateTime.now.iso8601,
        tests: tests,
        env: {
          source: 'travis',
          info: [
            'CPU: ' + 'TODO',
            'Platform: ' + RUBY_PLATFORM,
            'Build id: ' + env('TRAVIS_BUILD_ID', default: 'undefined'),
            'Job id: ' + env('TRAVIS_JOB_ID', default: 'undefined')
          ].join('\n')
        },
      },
    })
  end

  def self.compacterize(obj)
    case obj
      when Array then obj.compact.map { |item| compacterize(item) }
      when Hash then obj.each_with_object({}) { |(key, item), hash| !item.nil? && hash[key] = compacterize(item) }
      else obj
    end
  end

  def self.env(name, options)
    if ENV[name] && ENV[name] != options[:ignore]
      ENV[name]
    else
      options[:default]
    end
  end
end

if __FILE__ == $0
  file = ARGV[0]
  addr = ARGV[1]

  if !file || !addr
    $stderr.puts "usage: ruby #{__FILE__} tests.json http://shelfgauge.url"
    exit 1
  end

  tests = JSON.parse(IO.read(file))

  ShelfgaugeReport.post(tests, addr) do |body|
    puts body
  end
end
